snapcraft/application.py

```diff
@@ -22,87 +22,129 @@
 import pathlib
 import signal
 import sys
-from typing import Any, List, cast
+from typing import Any, List
 
 import craft_application.commands as craft_app_commands
 import craft_application.models
 import craft_cli
 import pydantic
 from craft_application import Application, AppMetadata, util
+from craft_application.errors import CraftValidationError
 from craft_application.models import BuildInfo
 from craft_cli import emit
 from craft_providers import bases
 from overrides import override
 
 from snapcraft import cli, commands, errors, models, services
 from snapcraft.commands import unimplemented
+from snapcraft.const import SUPPORTED_ARCHS
 from snapcraft.extensions import apply_extensions
-from snapcraft.models import Architecture
-from snapcraft.models.project import apply_root_packages, validate_architectures
+from snapcraft.models import Platform
+from snapcraft.models.project import apply_root_packages
 from snapcraft.providers import SNAPCRAFT_BASE_TO_PROVIDER_BASE
 from snapcraft.utils import get_effective_base, get_host_architecture
 
 
 class SnapcraftBuildPlanner(craft_application.models.BuildPlanner):
     """A project model that creates build plans."""
 
-    architectures: List[str | Architecture] = pydantic.Field(
-        default_factory=lambda: [get_host_architecture()]
-    )
-    base: str | None = None
+    base: str | None
     build_base: str | None = None
-    name: str | None = None
+    name: str
+    platforms: dict[str, Any]
     project_type: str | None = pydantic.Field(default=None, alias="type")
 
-    @pydantic.validator("architectures", always=True)
-    def _validate_architecture_data(  # pylint: disable=no-self-argument
-        cls, architectures: list[str | Architecture]
-    ) -> list[Architecture]:
-        """Validate architecture data.
-
-        Most importantly, converts architecture strings into Architecture objects.
-        """
-        return validate_architectures(architectures)
-
-    def get_build_plan(self) -> List[BuildInfo]:
-        """Get the build plan for this project."""
-        build_plan: List[BuildInfo] = []
+    @pydantic.validator("platforms")
+    @classmethod
+    def _validate_all_platforms(cls, platforms: dict[str, Any]) -> dict[str, Any]:
+        """Make sure all provided platforms are tangible and sane."""
+        for platform_label in platforms:
+            platform: dict[str, Any] = (
+                platforms[platform_label] if platforms[platform_label] else {}
+            )
+            error_prefix = f"Error for platform entry '{platform_label}'"
+
+            # Make sure the provided platform_set is valid
+            try:
+                platform = Platform(**platform).dict()
```

Comment:
Out of curiosity why are you converting this back to a dict? It doesn't seem like you're doing anything that couldn't be done with `platform.build_on` and `platform.build_for`?

Reply:
Fixed